### PR TITLE
8273544: Increase test coverage for snippets

### PR DESCRIPTION
**Pull request by pavelrappo**
This change is mostly about tests. I say "mostly" because while increasing code coverage, which was the primary goal of this exercise, I uncovered a few non-critical bugs and fixed them in-place. The net effect of the change boils down to these code coverage statistics.

before
------

%method	%block	%branch	%line
jdk.javadoc.internal.doclets.toolkit.taglets.SnippetTaglet	
75%(12/16)
87%(109/124)
88%(99/112)
85%(140/164)

#classes	%method	%block	%branch	%line
jdk.javadoc.internal.doclets.toolkit.taglets.snippet	
70%(80/114)
76%(310/407)
65%(178/273)
81%(413/508)

after
-----

%method	%block	%branch	%line
jdk.javadoc.internal.doclets.toolkit.taglets.SnippetTaglet	
100%(17/17)
95%(120/126)
93%(103/110)
97%(163/168)

%method	%block	%branch	%line
jdk.javadoc.internal.doclets.toolkit.taglets.snippet	
83%(94/112)
85%(348/405)
73%(202/273)
91%(463/505)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8273544](https://bugs.openjdk.java.net/browse/JDK-8273544): Increase test coverage for snippets

### Reviewers
 * [Jonathan Gibbons](https://openjdk.java.net/census#jjg) (@jonathan-gibbons - **Reviewer**) ⚠️ Review applies to 981cc305a36c0e1b3f4c11d14a393f5e8ae90671

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6359/head:pull/6359` \
`$ git checkout pull/6359`

Update a local copy of the PR: \
`$ git checkout pull/6359` \
`$ git pull https://git.openjdk.java.net/jdk pull/6359/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6359`

View PR using the GUI difftool: \
`$ git pr show -t 6359`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6359.diff">https://git.openjdk.java.net/jdk/pull/6359.diff</a>

</details>

### CODE REVIEW

File: test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetTag.java
Comment (hns) [inline test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetTag.java:71]: I assume the purpose of this naming scheme is to make visible what combinations of features are covered by each test. I'm not sure I would consider this enough benefit to justify method names which are very verbose and hard to read (especially when you don't have fresh memory of the scheme above).

Comment (pavelrappo) [reply test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetTag.java:71]: When one is planning to have many methods, one should better have a plan for how to maintain them. Such a plan should cover navigation, search, and if growth is foreseen, naming new methods. A naming scheme could in itself be such a plan.

I've tried alternative naming schemes but found this one to work best. It's obviously possible that I missed something. If you have other ideas, please share and let's discuss them.

Comment (pavelrappo) [reply test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetTag.java:71]: > I assume the purpose of this naming scheme is to make visible what combinations of features are covered by each test. I'm not sure I would consider this enough benefit to justify method names which are very verbose and hard to read (especially when you don't have fresh memory of the scheme above).

You are right when saying that this naming scheme helps to survey the test landscape. This is especially helpful while mass-developing tests, which is what I'm doing at the moment.

Comment (pavelrappo) [reply test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetTag.java:71]: Do you think we could add an underscore here and there to improve readability? For example, we could separate the first four parts of the name from the last one, which is a custom string:
```
testNegativeExternalTag_FileNotFoundModuleSourcePath
```
I'm not against "camelCase", but I have to admit that "snake_case" or a reasonable mix of both styles might work better for longer names.

File: test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetPathOption.java
Comment (jonathan-gibbons) [inline test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetPathOption.java:46]: (minor) move to SnippetTester

File: test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetPathOption.java
Comment (jonathan-gibbons) [inline test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetPathOption.java:131]: (minor) This sequence is showing up enough to warrant being a method somewhere.

File: test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetPathOption.java
Comment (jonathan-gibbons) [inline test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetPathOption.java:168]: Not wrong, but arguably somewhat superfluous. This is close to being a file manager test, not a snippet test.

File: test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetTag.java
Comment (jonathan-gibbons) [inline test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetTag.java:1193]: (style) why not indent the text?

File: test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetTag.java
Comment (jonathan-gibbons) [reply test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetTag.java:71]: In general, given the size of the file, I like the verbose names, if only because it gives the reader a strong hint up front about the form (nature) of the test.

That being said, now that you have introduced `SnippetTester` (which is great!) you could split this file. I would definitely group positive and negative test cases in the same file, but you could maybe separate `Tag` and `Markup` tests into separate files. (Just a suggestion).   

FWIW, my experience is that it is easier to test/debug files with fewer methods in, although it is on my to-do list to make it easier to run individual test cases.

**Review by jonathan-gibbons — COMMENTED** [review]
In general, great work.    Most of my comments are about style, especially unindented text blocks.

The one aspect that I think deserves attention is the code for the `@link` feature and the use of in-memory file objects and file managers. While not wrong, it seems overkill and unnecessary, since I think you could embed the reference tag and the markup tag in the same file, without going all the way to the separate run of javadoc.   Given that it's not wrong, and that there are dependent reviews blocked by this, I'll approve it if you want, but would like to understand if there are reasons to do it as you have.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/SnippetTaglet.java
Comment (pavelrappo) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/SnippetTaglet.java:202]: You are much more knowledgeable in FileManager than I am. My thinking was that while it's probably not an issue with the default file manager, I couldn't guarantee that IllegalArgumentException is not thrown if this code runs with a custom file manager.

File: test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetMarkup.java
Comment (pavelrappo) [reply test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetMarkup.java:71]: I was surprised to hear that this is a convention, so much so that I wrote a script to survey the JDK tests in this regard.
The script showed that while what you said is true for javadoc, it is not so true for other areas (even in langtools). That said, I can see the rationale for having this convention and will update the snippet tests accordingly.

Separately. If there's a value in mass-fixing the JDK tests so that they follow this convention, I can create a PR to do that. Here are some numbers provided by the script, to assess the situation. There are 32,920 `test/**/*.java` files. Of these, 19,190 have jtreg comments. Of these, 4,284 (or 22%) do not follow the convention one way or another. If pursued, that might result in an extremely tedious-to-review PR.

File: test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetMarkup.java
Comment (pavelrappo) [reply test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetMarkup.java:182]: Thanks. It is mainly for readability, to clearly show how the `@link` is expected to work. That said, the construct is somewhat clumsy. I hope to improve it, once templated strings arrive (https://openjdk.java.net/jeps/8273943).

File: test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetMarkup.java
Comment (jonathan-gibbons) [reply test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetMarkup.java]: Maybe the convention is not as strong as I thought. It definitely does not warrant mass updates.

I'd be OK to review any updates needed for javadoc tests.

File: test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetMarkup.java
Comment (pavelrappo) [reply test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetMarkup.java:385]: I thought about it and even half-implemented it before creating this PR. But a test purist in me demanded to have a clear way to get the link without interfering with the test logic. If it is not a showstopper, I would really like to integrate this PR to unlock more urgent work on the snippets feature before RDP 1, which is on 2021/12/09 as per the Schedule at https://openjdk.java.net/projects/jdk/18/. Sometime after this PR is integrated, we can revisit this.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/SnippetTaglet.java
Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/SnippetTaglet.java:202]: I was just reading the documented conditions for this `JavaFileManager` method to throw `IllegalArgumentException`. That being said, one of the problems with ILA, and potentially other unchecked exception, is that they can occur from deep within the method that is called. 

FYI, javac has a way of wrapping user-provided components so that it can better handle unexpected exceptions from user-provided code as compared to unexpected exceptions in javac itself. If you're interested, see`javac` `ClientCodeWrapper`.   And, to be clear, I'm *not* suggesting that for javadoc at this time!

File: test/langtools/jdk/javadoc/doclet/testSnippetTag/SnippetTester.java
Comment (jonathan-gibbons) [inline test/langtools/jdk/javadoc/doclet/testSnippetTag/SnippetTester.java:96]: This works, and is a clever solution for now.  Maybe "one of these days" we remove the `checkOutputEither` method.

File: test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetTag.java
Comment (jonathan-gibbons) [inline test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetTag.java:98]: General comment, applying to all such situations ... I like the latest iteration on the naming scheme, with the embedded `_`.

File: test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetMarkup.java
Comment (pavelrappo) [reply test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetMarkup.java:385]: What you suggested is how I had that initially. But a test purist in me demanded a clear way of getting the link without interfering with the test logic.

As for text blocks: to me, indenting a text block is a complex, subjective and unsettled field. I would postpone re-indenting them for now, primarily because of the churn it would otherwise introduce to the dependent PRs. 

If neither of these issues is a showstopper, I would really like to integrate this PR to unlock more urgent work on the snippets feature before RDP 1, which is on 2021/12/09 as per the Schedule at https://openjdk.java.net/projects/jdk/18/. Sometime after this PR is integrated, we can revisit this.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/snippet/Attribute.java
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/snippet/Attribute.java:61]: General question:   the commit message is:

> So we can finally use sealed classes where they were originally envisioned.

But, the code edits seem to be backing away from using sealed classes and interfaces.  What happened?

File: test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetPathOption.java
Comment (pavelrappo) [reply test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetPathOption.java:163]: I can add a comment that this test is a part of an end-to-end demonstration of how --snippet-path should work. This is a black-box test that assumes nothing about the implementation of the feature.

File: test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetMarkup.java
Comment (jonathan-gibbons) [reply test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetMarkup.java:385]: OK, but the test purist in me dislikes the significant amount of inline overhead to address this concern.

I'll pass on this for now, because I agree with unblocking the dependent PRs, but it seems like that is a negative side-effect of dependent PRs in general.  Once we get past this, at a minimum, I would suggest moving the in-memory file manager out of this class, and then potentially reconsider the decision that it is good practice to do the separate run.

Comment (jonathan-gibbons) [reply test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetMarkup.java:385]: > As for text blocks: to me, indenting a text block is a complex, subjective and unsettled issue. I would postpone re-indenting them for now, primarily because of the churn it would otherwise introduce to the dependent PRs.

_complex, subjective and unsettled_?  Really? JEP 378 would indicate otherwise!   So, I disagree with the lack of indent, but (reluctantly) agree on the churn causing issues for dependent t PRs.

**Review by jonathan-gibbons — APPROVED** [review]
While I disagree with some of the code here, the code is not wrong, and so I'll approve the review to unblock the dependent work.

I hope we can clean up some of the issues later on.

Setting those concerns aside, overall, this is great work. Well done for such a thorough job for improving test coverage.

File: test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetPathOption.java
Comment (jonathan-gibbons) [reply test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetPathOption.java:163]: OK, but I still think the test is somewhat unnecessary. But now you have it, don't delete it.

File: test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetMarkup.java
Comment (jonathan-gibbons) [reply test/langtools/jdk/javadoc/doclet/testSnippetTag/TestSnippetMarkup.java:385]: (Cleanup suggestion, for later?)  maybe embed the `{@link}` in a separate comment, maybe even in a separate class or file, but still processed in the same run of `javadoc`.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/SnippetTaglet.java
Comment (pavelrappo) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/SnippetTaglet.java:202]: > That being said, one of the problems with ILA, and potentially other unchecked exception, is that they can occur from deep within the method that is called.

Care to elaborate?

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/snippet/Attribute.java
Comment (pavelrappo) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/snippet/Attribute.java:61]: This is embarrassing. The commit clearly contradicts its message, and I don't immediately understand why. I'll try to figure out it before integrating. Thanks.

Comment (pavelrappo) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/snippet/Attribute.java:61]: I vaguely remember that at some stage I seem to have realized that while `sealed` works for Style, it does not work for Attribute. This is because the attribute is typically queried by the most specific `Class<T extends Attribute>`, which means that the client knows the exact type of the attribute upon return from Attributes.get (note plural attributeS).

Here's another issue that might have contributed to this mess: although the [sealed classes](https://openjdk.java.net/jeps/409) are available, the [exhaustive switch for sealed hierarchies](https://openjdk.java.net/jeps/420) is not.

I will push a commit to make Style `sealed`, which seems reasonable.

Comment (pavelrappo) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/snippet/Attribute.java:61]: I vaguely remember that at some stage I seem to have realized that while `sealed` works for Style, it does not work for Attribute. This is because the attribute is typically queried by the most specific `Class<T extends Attribute>`, which means that the client knows the exact type of the attribute upon return from Attributes.get (note plural attributeS).

Here's another issue that might have contributed to this mess: although the [sealed classes](https://openjdk.java.net/jeps/409) are available, the [exhaustive switch for sealed hierarchies](https://openjdk.java.net/jeps/420) is not.

That said, here's what I will do. I will push a commit to make Style `sealed`, which a seems reasonable thing to do. Do you have any objections, Jon? (Asking because you have already reviewed the change.)

Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/snippet/Attribute.java:61]: Thanks for the explanation.   I have no strong preferences either way, with a mild preference for leaving it alone, and improve it later when you may be able to make better use of the declaration.

Comment (pavelrappo) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/snippet/Attribute.java:61]: Please clarify: do you want me to revert 93f01b9, push the PR as it is now, or do something else?

Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/snippet/Attribute.java:61]: by défault, push the PR as is, unless you have strong feelings to do otherwise.

Comment (jonathan-gibbons) [reply src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/snippet/Attribute.java:61]: >> I will push a commit to make Style sealed, which seems a reasonable thing to do. Do you have any objections, Jon?

No objections to that course of action.

**Pull request closed without merging**
